core/arch/x86/encode.c

```diff
@@ -844,6 +853,12 @@ reg_size_ok(decode_info_t *di /*prefixes field is IN/OUT; x86_mode is IN*/, reg_
         }
         return true;
     }
+    if (optype == TYPE_K_REG || optype == TYPE_K_MODRM || optype == TYPE_K_MODRM_R ||
+        optype == TYPE_K_VEX) {
+        return (opsize == OPSZ_1 || opsize == OPSZ_2 || opsize == OPSZ_4 ||
+                opsize == OPSZ_8);
```

Comment:
Won't the size_ok() call right above handle these?  What is the reason they need special code?

Reply:
The existing code make its decision based on the operand and prefixes. The special thing about these scalar k register instructions is that the k register size is implied by its opcode (w, b, q, d versions). One could move this into the size_ok function, but that would simply be a re-factor. I think outside makes more sense.

Reply:
There must be precedent for this: does it really need special handling differently from other registers?  Various SIMD opcodes operate on different parts of the *mm registers as specified by the opcode and we don't have special code here for them.  Maybe these mask registers should be treated like the sub-pieces of the *mm registers.  (The other way is to give the sub-registers separate names: like al, ax, eax, rax.)

Reply:
This still doesn't seem necessary: size_ok collapses sub-reg sizes.  Are the smaller sizes sub-reg sizes, just like pieces of xmm registers?  It seems like they are, and thus size_ok will work fine, if we make them OPSZ_{1,2,4}_of_8.